src/tsd-metrics.ts

```diff
@@ -0,0 +1,359 @@
+import timerSamples = require("./timer-samples");
+import counterSamples = require("./counter-samples");
+import units = require("./tsd-units");
+import sample = require("./tsd-metric-sample");
+import metricsList = require("./tsd-metrics-list");
+import metricsEvent = require("./tsd-metrics-event");
+import metricsFactory = require("./tsd-metrics-factory");
+import log4jsSink = require("./sinks/tsd-query-log-sink");
+import consoleSink = require("./sinks/tsd-console-sink");
+import sink = require("./sinks/tsd-sink");
+import errors = require("./error-reporting");
+
+//aliases
+import TsdTimer = timers.TsdTimer;
+import TsdCounter = counters.TsdCounter;
+import Lazy = utils.Lazy;
+import CounterSamples = counterSamples.CounterSamples;
+import TimerSamples = timerSamples.TimerSamples;
+import Units = units.Units;
+import TsdMetricSample = sample.TsdMetricSample;
+import TsdMetricsList = metricsList.TsdMetricsList;
+import TsdMetricsEvent = metricsEvent.TsdMetricsEvent;
+import TsdMetricsFactory = metricsFactory.TsdMetricsFactory
+
+/**
+ * Node JS class for publishing metrics as time series data (TSD).
+ *
+ * @class
+ * @author Mohammed Kamel (mkamel at groupon dot com)
+ */
+export class TsdMetrics implements tsdDef.Metrics {
+
+    /**
+     * Metrics state object that holds flag to the metrics open state along with common event emitter.
+     */
+    private _metricsStateObject:MetricsStateObject = new MetricsStateObject();
+    private _timers:{[name:string]: tsdDef.Timer} = {};
+    private _counters:{[name:string]: tsdDef.Counter} = {};
+    private _sinks:tsdDef.Sink[];
+
+    /**
+     * MetricsEvent field that will be used to serialize the final metric event.
+     */
+    private _metricsEvent:tsdDef.MetricsEvent = new TsdMetricsEvent();
+
+    private static getOrCreateMetricsCollection<T>(name:string, collectionOfT:{[name:string]:T}, factory:() => T) {
+        if (collectionOfT[name] === undefined) {
+            collectionOfT[name] = factory();
+        }
+        return collectionOfT[name];
+    }
+
+    private getOrCreateCounterSamples(name:string):CounterSamples {
+        return <CounterSamples>TsdMetrics.getOrCreateMetricsCollection(
+            name, this._metricsEvent.counters, () => new CounterSamples(name, this._metricsStateObject));
+    }
+
+    private getOrCreateTimerSamples(name:string):TimerSamples {
+        return <TimerSamples>TsdMetrics.getOrCreateMetricsCollection(
+            name, this._metricsEvent.timers, () => new TimerSamples(name, this._metricsStateObject));
+    }
+
+    private getOrCreateGaugesSamples(name:string):tsdDef.MetricsList < tsdDef.MetricSample > {
+        return TsdMetrics.getOrCreateMetricsCollection(
+            name, this._metricsEvent.gauges, () => new TsdMetricsList < tsdDef.MetricSample >());
+    }
+
+    /**
+     * Constructor.
+     */
+    constructor(serviceName:string, clusterName:string, hostNameSupplier:tsdDef.HostNameSupplier, sinks:tsdDef.Sink[]) {
```

Comment:
The host name should be realized at the time of metrics instance construction. Leave the supplier in the factory.